.github/scripts/snapshot_images.sh

```diff
@@ -0,0 +1,83 @@
+#!/bin/bash
+
+# Initialize array to store digest SHAs
+declare -a digests
+
+# Fetch all docker image names
+image_names=$(gh api \
+  -H "Accept: application/vnd.github+json" \
+  -H "-GitHub-Api-Version: 2022-11-28" \
+  -H "Authorization: Token ${GITHUB_TOKEN}" \
+  --paginate "/orgs/boozallen/packages?package_type=container" | jq -r '.[] | select((.name | startswith("aissemble")) and (.name | endswith("-chart") | not)) | .name')
+
+# For each docker image, find their snapshot versions by grabbing any tag that ends with "-SNAPSHOT"
+for name in $image_names; do
+  all_snapshot_versions=$(gh api \
+    -H "Accept: application/vnd.github+json" \
+    -H "-GitHub-Api-Version: 2022-11-28" \
+    -H "Authorization: Token ${GITHUB_TOKEN}" \
+    --paginate "/orgs/boozallen/packages/container/${name}/versions" \
+    | jq -r '.[] | .metadata.container.tags[] | select(endswith("-SNAPSHOT"))'| jq -R -s 'split("\n")| map(select(length > 0))')
+
+  # Find the latest snapshot version by sorting all snapshot versions, then selecting the top
+  latest_snapshot_version=$(echo "$all_snapshot_versions" | jq -r '.[] | select(endswith("-SNAPSHOT"))' | sort -r | head -n 1)
+
+  echo "Processing snapshot image ${name}:${latest_snapshot_version}"
+
+  # Fetch the latest snapshot version SHA
+  latest_snapshot_manifest_list_shas=$(docker buildx imagetools inspect --raw "ghcr.io/boozallen/${name}:${latest_snapshot_version}" | jq -r '.manifests[].digest' | paste -s -d ' ' -)
+
+  echo "Manifest List: $latest_snapshot_manifest_list_shas"
+
+  # Add to digests array
+  digests+=("$latest_snapshot_manifest_list_shas")
+
+  # Find if there are any patch versions available
+  # Extract version from latest snapshot
+  version_part="${latest_snapshot_version%-SNAPSHOT}"
+
+  # Extract the major, minor, and patch components using IFS (Internal Field Separator)
+  IFS='.' read -r major minor patch <<< "$version_part"
+  previous_major=$((major - 1))
+  previous_minor=$((minor - 1))
+
+  # If previous_minor is >= 0, then patch pattern = [current_major].[previous_minor].[1-9]-SNAPSHOT
+  if [ $previous_minor -ge 0 ];then
+    patch_major=${major}
+    patch_minor=${previous_minor}
+  else
+   # Else previous_minor is < 0, then patch pattern = [previous_major].[10 + previous_minor].[1-9]-SNAPSHOT
+    patch_major=${previous_major}
+    patch_minor=$((10 + previous_minor))
+  fi
+
+  patch_pattern="${patch_major}\.${patch_minor}\.[1-9]-SNAPSHOT"
+
+  for version in $(echo "$all_snapshot_versions" | jq -r '.[]'); do
+    if [[ ${version} =~ ^${patch_pattern}$ ]]; then
+      echo "Patch version ${version} matches patch pattern ${patch_pattern}"
+
+      # Add to array containing all matching patch versions
+      matching_patch_versions+=("$version")
+    fi
+
+    # If matching patch versions array is not empty
```

Comment:
I: the logic to find the latest patch version is as follows:

- parse the latest snapshot version (follows `major.minor.0-SNAPSHOT` syntax) to find the `major` and `minor` versions
- find the previous minor version
- if the previous minor is greater than or equal to 0, then the patch version uses the same major version and the pattern to look for is `current_major.previous_minor.[1-9]-SNAPSHOT`
- if the previous minor version is less than 0, we need to also find the previous major version and the patch version pattern to check for becomes `previous_major.(10 + previous_minor).[1-9]-SNAPSHOT`
- save all found patch versions to an array
- if the array is not empty, sort the contents, find the latest patch version, and pull the SHA for that `image:latest_patch_version` and save to main `digests` array


Since there are no actual patch versions currently, this logic was tested locally by creating dummy data. Some examples below:
```
All Snapshot Versions:
[ "1.9.0-SNAPSHOT", "1.8.3-SNAPSHOT", "1.8.1-SNAPSHOT", "1.7.0-SNAPSHOT", "1.8.0-SNAPSHOT", "1.7.1-SNAPSHOT" ]
Latest Snapshot Version found: 1.9.0-SNAPSHOT
Previous major: 0
Previous minor: 8
Previous minor is greater than or equal to 0
Using Patch Major: 1 and Patch Minor: 8
Patch Pattern: 1\.8\.[1-9]-SNAPSHOT

Patch version 1.8.3-SNAPSHOT matches patch pattern
Patch version 1.8.1-SNAPSHOT matches patch pattern
Sorted matching patch versions: 1.8.3-SNAPSHOT 1.8.1-SNAPSHOT
Latest patch version found: 1.8.3-SNAPSHOT
```
```
All Snapshot Versions:
[ "1.9.0-SNAPSHOT", "1.8.3-SNAPSHOT", "1.8.1-SNAPSHOT", "2.0.0-SNAPSHOT", "1.9.1-SNAPSHOT", "1.9.4-SNAPSHOT" ]
Latest Snapshot Version found: 2.0.0-SNAPSHOT
Previous major: 1
Previous minor: -1
Previous minor is less than 0
Using Patch Major: 1 and Patch Minor: 9
Patch Pattern: 1\.9\.[1-9]-SNAPSHOT

Patch version 1.9.1-SNAPSHOT matches patch pattern
Patch version 1.9.4-SNAPSHOT matches patch pattern
Sorted matching patch versions: 1.9.4-SNAPSHOT 1.9.1-SNAPSHOT
Latest patch version found: 1.9.4-SNAPSHOT
```

Reply:
I think there's a false assumption in this logic that the major.minor.patch number has to be 1-9 which is not always the case. Before we released aissemble `1.0.0` we were on version `0.13.6`.

The logic should still be fine for when the major remains the same, but probably won't work for when the major changes.

Reply:
Yeah you're right, I was assuming that our versioning would be 1-9 and sequential... 

I think it make more sense if I just use the latest release version's major and minor values for the patch pattern. 
For example, since the latest release is `1.8.0`, the patch pattern would be `1\.8\.[1-9]+[0-9]*-SNAPSHOT?`. What do you think?

Reply:
I agree. Honestly it's not the end of the world to be preserving these patch release snapshots anyway. The one that really matters is the patch release itself.

Reply:
@carter-cundiff updated to new logic of using the latest release version's major and minor versions for the patch pattern. The latest release version is passed from the `Fetch multi-platform package version SHAs` step into `GITHUB_OUTPUT` and then used in the `Fetch latest snapshot version SHAs` step.

Reply:
Don't think this works for when the latest release is `1.0.1` and we're working on a patch release `1.0.2`. Should probably just match on whatever the current patch it + 1 instead of `.[1-9]+[0-9]*-SNAPSHOT?"`

Reply:
I guess it would match on both, but it this scenario we want it to match on only `1.0.2-SNAPSHOT`